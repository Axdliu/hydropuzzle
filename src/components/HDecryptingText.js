import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import ReactTimeout from 'react-timeout';
import shuffle from 'shuffle-array';

import HText from './HText';

/**
 * Decrypting text effect
 *
 * Inspired by https://formidable.com/about/ (hover effect for github)
 */
class HDecryptingText extends PureComponent {
  static propTypes = {
    completed: PropTypes.bool,
    onEnd: PropTypes.func,
    text: PropTypes.string.isRequired
  }

  static defaultProps = {
    completed: false,
    onEnd: () => {}
  }

  delay = 5;
  text = this.props.text.trim();
  charsIndex = []

  state = {
    text: this.text.replace(/[^\s]/g, '*')
  }

  constructor (props) {
    super(props);

    // SMELL: Array.from(Array(this.text.length).keys());
    //        somehow doesn't work on... Android
    for (let i = 0; i < this.text.length; i += 1) {
      this.charsIndex.push(i);
    }
    shuffle(this.charsIndex);
  }

  componentDidMount () {
    if (this.props.completed) {
      this.setState({text: this.text});
      return;
    }

    this.props.setTimeout(this.nextLetter, 1000);
  }

  render () {
    return (
      <HText type='mono' {...this.props}>
        {this.state.text}
      </HText>
    );
  }

  nextLetter = () => {
    const index = this.charsIndex.shift();
    const text = this.state.text.slice(0, index) + this.text[index] + this.state.text.slice(index + 1);

    this.setState({ text });

    if (this.charsIndex.length === 0) {
      this.props.setTimeout(this.props.onEnd, 500);
      return;
    }

    this.props.setTimeout(this.nextLetter, this.delay);
  }
}

export default ReactTimeout(HDecryptingText);
