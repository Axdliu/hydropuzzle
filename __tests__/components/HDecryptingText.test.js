import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
// import { shallow } from 'enzyme';

import HDecryptingText from '@components/HDecryptingText';

const text = 'Encrypted text!';

describe('HDecryptingText', () => {
  const onEnd = jest.fn();
  const component = (
    <HDecryptingText
      text={text}
      onEnd={onEnd} />
  );

  it('renders correctly', () => {
    const wrapper = renderer.create(component);

    jest.runAllTimers();

    expect(wrapper).toMatchSnapshot();
    expect(onEnd).toBeCalled();
  });
});
