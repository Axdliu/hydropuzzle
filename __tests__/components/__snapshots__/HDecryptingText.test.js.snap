// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HDecryptingText renders correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  cancelAnimationFrame={[Function]}
  clearImmediate={[Function]}
  clearInterval={[Function]}
  clearTimeout={[Function]}
  completed={false}
  ellipsizeMode="tail"
  onEnd={[Function]}
  requestAnimationFrame={[Function]}
  setImmediate={[Function]}
  setInterval={[Function]}
  setTimeout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "color": "black",
        "fontSize": 17.05,
        "fontWeight": "400",
      },
      Object {
        "fontFamily": "Courier",
        "lineHeight": 23.25,
      },
      Object {},
    ]
  }
  text="Encrypted text!"
  type="mono"
>
  Encrypted text!
</Text>
`;
